zebra-test/src/service_extensions.rs

```diff
@@ -0,0 +1,26 @@
+//! Extension traits for [`Service`] types to help with testing.
+
+use futures::FutureExt;
+use tower::{Service, ServiceExt};
+
+/// An extension trait to check if a [`Service`] is immediately ready to be called.
+pub trait IsReady<Request>: Service<Request> {
```

Comment:
Undocumented Behaviour: this convenience wrapper hides a call to `now_or_never`, which never registers the task for waking. So after these methods are called once on a service, any further calls to the service may produce incorrect test results.

So we can either:
1. Revert this change, or
2. Make the new wrapper harder to misuse:
    * add `OrNever` to the wrapper name
    * modify its methods to take ownership of the service
    * summarise the `now_or_never` behaviour in the docs

It would also be nice to prevent the wrapper methods being called on a cloned, buffered service. But our only way of doing that is educating developers. We can teach everyone about the wrapper, or revert the wrapper, and just teach about `now_or_never`.

We can't switch to a `now_or_never` alternative that can be called multiple times. Because we can't obtain service readiness multiple times anyway. (Readiness must be used after it is obtained, particularly for buffered services.)

Reply:
This is a valid concern, and I thought it might be a good idea to use `NowOrLater` instead. To do so, I had to move the `NowOrLater` type from `zebrad` into a new `now-or-later` crate in order to avoid cyclic dependencies.

Reply:
Thanks for this fix!

How can we prevent multiple service readiness calls in the tests?

> We can't switch to a now_or_never alternative that can be called multiple times. Because we can't obtain service readiness multiple times anyway. (Readiness must be used after it is obtained, particularly for buffered services.)

If the service isn't buffered, and it has no internal reservations, it is ok to check readiness multiple times:
https://docs.rs/tower/0.4.11/tower/trait.Service.html#tymethod.poll_ready

But if the service makes internal reservations, or it is buffered, or it checks readiness for buffered services, we can't call it multiple times:
https://docs.rs/tower/0.4.11/tower/buffer/struct.Buffer.html#a-note-on-choosing-a-bound

I'd like to take ownership of the service, if we can. And I'd like to document this issue in `IsReady`.

Reply:
This seems like non-trivial work. Could you open an issue for us to do it later?

Reply:
Since `mpsc` channels have capacity limits, this could be a real bug in this refactor.

Allowing the service to be re-used after `is_ready` could hide test regressions in this PR. So I'd really like to do it as part of this PR, to make sure the refactor works correctly.

The previous code took ownership of the `ServiceExt::ready` future, but didn't take ownership of the service. So the bug could still have happened then. (But I was very careful to avoid it.)

I would like to make it easier to maintain this code, and avoid the risk of subtle bugs in this PR, or in future refactors. I'll see if I can fix it today.

Reply:
> Could you open an issue for us to do it later?

See PR #3262, which takes ownership of the service, and documents this issue in `IsReady`.

I also opened ticket #3261 for some larger API changes.

Reply:
Here's an example of a potential bug discovered by this fix:

> assertion failed: harness.current_error().is_none()

in `client_service_ready_ok`:

```rust
    assert!(client.is_ready().await);
    assert!(harness.current_error().is_none());
    assert!(harness.wants_connection_heartbeats());
    assert!(harness.try_to_receive_outbound_client_request().is_empty());
```

https://github.com/ZcashFoundation/zebra/runs/4577397601?check_suite_focus=true#step:14:984

`client.is_ready()` might reserve capacity in the mpsc channel (it's undocumented), and the channel is small, so we need to document why the test is valid. Or re-arrange the test so it doesn't matter.

I covered both scenarios in PR #3262 by adding a new test (just in case the behaviour is ever different).